Add verifying key and proving key serialization

jonathanpwang opened the pull request:
This PR adds features for reading and writing the verifying key and proving key of any circuit from/to file. 

For verifying key, this uses https://github.com/zcash/halo2/pull/661 which fixes the issue with vkey write not working after selector compression. 

For proving key, this uses serde derive macros and `bincode`. 

Both vkey and pkey serializations have been tested on very large circuits (up to `K = 23`), including circuits created from `plonk-verifier`. There have been no problems even up to pkey of sizes `> 100GB`.

han0110 commented on halo2_proofs/src/plonk.rs:
```diff
@@ -159,6 +229,44 @@ impl<C: CurveAffine> ProvingKey<C> {
+            permutation: self.permutation.clone(),
+            ev: self.ev.clone(),
+        };
+        bincode::serialize_into(writer, &partial_pkey).expect("should be able to serialize pkey");
```
I think we can add another variant in `plonk::Error` for pk/vk serde error (even just `Serde(err)` should be better than panic).

han0110 commented on halo2_proofs/src/plonk.rs:
```diff
@@ -159,6 +229,44 @@ impl<C: CurveAffine> ProvingKey<C> {
+    ) -> io::Result<Self> {
+        let vk = VerifyingKey::<C>::read::<R, ConcreteCircuit>(reader, params)?;
+        let partial_pk: ProvingKeyWithoutVerifyingKey<C> =
+            bincode::deserialize_from(reader).expect("should be able to deserialize pkey");
```
ditto

han0110 reviewed (commented):
Overall look good to me! Just with some more proper error propagation I think it's good to merge.

I feel like next we can try to serialize the `ConstraintSystem` with selector compressed in `VerifyingKey`, then we don't need to store selector values in vk and the verifier doesn't even need to know the concrete circuit but can still verify the proof.

jonathanpwang commented on halo2_proofs/src/plonk/evaluation.rs:
```diff
@@ -1,3 +1,4 @@
+use crate::helpers::SerdePrimeField;
```
I wasn't sure what you meant by this. Did you want to move changes from other files to this file?

han0110 replied:
Ah no, I meant we can "remove" all the changes in `evaluation.rs` since we don't need to serialize it anymore, sorry for confusing.

jonathanpwang commented:
Implemented @han0110 's suggestions.

CPerezz commented on halo2_proofs/src/plonk.rs:
```diff
@@ -159,6 +229,43 @@ impl<C: CurveAffine> ProvingKey<C> {
+            fixed_cosets: self.fixed_cosets.clone(),
+            permutation: self.permutation.clone(),
+        };
+        bincode::serialize_into(writer, &partial_pkey).map_err(Error::Serde)
```
Hardcoding `bincode` directly here is a bit tough TBH. But I understand that you can't use serde traits because that could allow yaml or anything esentially.

As said in my review comment, I'm a bit hessitant over the `serde`.

jonathanpwang replied:
Previously because of evaluation there was too much to implement `write` for directly, but since Han pointed out that we can recompute evaluation directly, it seems the only primitive necessary is to implement write for `Polynomial<F, _> = Vec<F>` and `Vec<Polynomial<F, _>>`, which seems doable. 

So if you want to limit dependencies, I think it is possible to just implement some `write` and `read` directly.

CPerezz replied:
I'd prefer it and avoid serde. Unless it is extremely painful.
If any of Han or Onur agree with me, I'd suggest to change it.

CPerezz reviewed (commented):
I'm a bit concerned about introducing all of the `serde` stack here. Why do we go for serde rather than a simpler `to/from_bytes` and maybe also a side `read/write` in case you prefer it??

Serde comes with A BUNCH of dependencies and it forces anything that is using this as a dependency to also use `serde`.

I would say that since the entire crate is done without using it, I'd follow the same rationale and structure for these structs. `read/write` and `to/from_bytes` might suit better.

I say that because it's annoying to use a lib as dependency (as halo2) and need to use serde just for 2 structures in it.

Any opinions @han0110 @kilic ?

CPerezz commented:
Hey @jonathanpwang  any news with that??

jonathanpwang commented:
> Hey @jonathanpwang any news with that??

Ah sorry I've been a bit busy. It may take a little longer (a week or two) before I get around to this, but just confirming I will try to do the serialization without `serde` soon.

jonathanpwang commented:
Okay I think I implemented the direct implementation of read/write by just serializing vectors in the obvious way. Also added an example for basic testing.

CPerezz commented on halo2_proofs/src/plonk.rs:
```diff
@@ -50,14 +51,82 @@ pub struct VerifyingKey<C: CurveAffine> {
+
+        // write self.selectors
+        for selector in &self.selectors {
+            let mut selector_bytes = vec![0u8; selector.len() / 8 + 1];
+            for (i, value) in selector.iter().enumerate() {
+                let byte_index = i / 8;
+                let bit_index = i % 8;
+                selector_bytes[byte_index] |= (*value as u8) << bit_index;
+            }
+            writer.write_all(&selector_bytes)?;
```
LGTM to compress the selectors into a single `u8`. I'd probably explain what is being don here with a quick and short comment to facilitate for people that aren't familiar.

CPerezz commented on halo2_proofs/src/plonk.rs:
```diff
@@ -50,14 +51,82 @@ pub struct VerifyingKey<C: CurveAffine> {
+                reader
+                    .read_exact(&mut selector_bytes)
+                    .expect("unable to read selector bytes");
+                for (i, value) in selector.iter_mut().enumerate() {
+                    let byte_index = i / 8;
+                    let bit_index = i % 8;
+                    *value = (selector_bytes[byte_index] >> bit_index) & 1 == 1;
```
You could create a `pack/unpack` fn pair that removes this chunk into another function. I think will simplify the read and write functions.

CPerezz commented on halo2_proofs/src/plonk/permutation.rs:
```diff
@@ -1,3 +1,5 @@
+use ff::PrimeField;
+
```
remove empty line

CPerezz commented on halo2_proofs/src/plonk/permutation.rs:
```diff
@@ -92,3 +109,50 @@ pub(crate) struct ProvingKey<C: CurveAffine> {
+    }
+}
+
+/// Reads a vector of polynomials from buffer
+pub(super) fn read_polynomial_vec<R: io::Read, F: PrimeField, B>(
+    reader: &mut R,
+) -> io::Result<Vec<Polynomial<F, B>>> {
+    let mut len_be_bytes = [0u8; 8];
+    reader.read_exact(&mut len_be_bytes)?;
+    let len = u64::from_be_bytes(len_be_bytes);
+
+    (0..len)
+        .map(|_| Polynomial::<F, B>::read(reader))
+        .collect::<io::Result<Vec<_>>>()
+}
+
+/// Writes a slice of polynomials to buffer
+pub(super) fn write_polynomial_slice<W: io::Write, F: PrimeField, B>(
+    slice: &[Polynomial<F, B>],
+    writer: &mut W,
+) -> io::Result<()> {
+    writer.write_all(&(slice.len() as u64).to_be_bytes())?;
+    for poly in slice.iter() {
+        poly.write(writer)?;
+    }
+    Ok(())
+}
```
I'd rather place these under `poly.rs` or rather `helpers.rs` than here. As this function has nothing to do with the permutation argument itself.

CPerezz commented on halo2_proofs/src/plonk/permutation.rs:
```diff
@@ -92,3 +109,50 @@ pub(crate) struct ProvingKey<C: CurveAffine> {
+pub(super) fn read_polynomial_vec<R: io::Read, F: PrimeField, B>(
+    reader: &mut R,
+) -> io::Result<Vec<Polynomial<F, B>>> {
+    let mut len_be_bytes = [0u8; 8];
+    reader.read_exact(&mut len_be_bytes)?;
+    let len = u64::from_be_bytes(len_be_bytes);
```
It'll never be possible of having a 2^64 term polynomial. I think `u32` is more than enough.

CPerezz reviewed (changes requested):
Thank you so much for the work on refactoring a bit this @jonathanpwang !

Just some comments and also 2 points I want to consider.
1. Could you also provide a `to/from_bytes` fn for ProverKey & VerifierKey??
2. The example you added feels extremely large for what we need to show which is how to use serde. I'm not even unsure if an example is needed TBH. But if we want it, I think we should use the smallest possible circuit and not have any comments for it. As at the end, we want to focus the example on the serialisation, not the circuit.

jonathanpwang commented:
I added `to/from_bytes` functions, though I was slightly lazy and just used the `Read` and `Write` traits for `&[u8]` and `Vec<u8>` respectively. For `to_bytes` I did optimize to avoid memory reallocation by having the vector's total capacity determined beforehand. 

Added `pack/unpack` functions as you suggested. 

Moved helper functions to `helper.rs`. 

Made the example smaller using a standard plonk example. I think it's still good to have the example just to have basic skeleton code for people to copy. Also it's helpful for testing.

CPerezz reviewed (approved):
LGTM!!!! Running the workflows now! 

Thanks for the work and the corrections @jonathanpwang :smile:

jonathanpwang commented:
I realized why reading the pkey was still kind of slow for >1GB proving keys: currently the `CurveAffine` implementation of `to_bytes` stores the point in compressed format, so when you read it you need to compute a cube and then square root to get the affine representation. 

Additionally, the `ff::PrimeField::to_repr` implementation stores the normal base representation of the field element, and when you read it it does a Montgomery conversion back to Montgomery form. 

I would prefer to just print out both `CurveAffine` and `PrimeField` in raw Montgomery form of the coordinates so that serialization/deserialization is only limited in speed by IO. This would double the size of the curve points, but the bulk of the proving key should be from the polynomials anyways. This would involve adding some traits + implementations such as `to_raw_bytes` and `from_raw_bytes_unchecked` to halo2curves. 

WDYT @CPerezz @han0110 ?

CPerezz commented:
I completelly agree with the idea @jonathanpwang @han0110 .
I've used this in `arkworks` a lot with `CanonicalSerialize` for example. Avoiding the checks while reading curve elements and giving up disk space in respect of faster reads.

Filled https://github.com/privacy-scaling-explorations/halo2curves/issues/9 to track this!

CPerezz commented:
Thanks @daira for the ping! <3
